### PR TITLE
[#192] Delete stream in frontend

### LINKED ISSUE #192: Delete a stream in the frontend
**Acceptance Criteria**
- For each stream in the view streams table, a delete button (such as a garbage can icon) is visible
- When I click on the delete button, a popup is shown "Are you sure you want to end the stream?"
- If the user clicks cancel / no, the popup closes and nothing happens
- If the user clicks yes, the delete stream endpoint is called, and the page in the frontend is updated to show that the stream doesn't exist anymore

### PR DESCRIPTION
**Pull request by gdzooks**
# General info

**Issue number**: 192

**Task description**: 
- For each stream in the view streams table, a delete button (such as a garbage can icon) is visible
- When the delete button is clicked, a popup is shown that reads "Are you sure you want to end the stream?"
- If the user clicks cancel, the popup closes and nothing happens
- If the user clicks yes, the delete stream endpoint is called, and the page in the frontend is updated to show that the stream doesn't exist anymore
 
 
| Original   | Updated|
|:---------------:|:---------------:|
| image | image |

# Testing
**Test coverage**:

* 100%! 
* Entire test suite for `StreamList` needs to be rewritten to use Enzyme but I don't think this is the issue/PR to do it in.

# Additional notes

**Note to reviewers**:

Couldn't get delete working with `SampleData` written the way it is - the whole class needs a big refactor if we want creating, modifying, and deleting to work with it (creating streams doesn't have a sample function either).
I don't have backend and/or samples setup so I didn't actually test that `axios.delete` worked as expected so I'm counting on one of you~

**To do before merging**:
N/A

### CODE REVIEW

File: frontend/src/api/StreamApi.js [hunk 1]
```diff
@@ -36,3 +36,12 @@ export function getStream(streamId) {
       .catch(reject);
   });
 }
+
+export function deleteStream(streamId) {
+  axios
+    .delete(`http://localhost:8080/stream/${streamId}`)
```
Comment (ButlerJAL) [inline frontend/src/api/StreamApi.js]: Change to `${process.env.REACT_APP_STREAM}/${streamId}`

File: frontend/src/streamlist/__tests__/StreamList.test.jsx [hunk 2]
```diff
@@ -92,3 +94,85 @@ test("Clicking dropdown on table row displays additional information", () => {
...
+  act(() => {
+    ReactTestUtils.Simulate.click(confirmButton);
+  });
+
+  // try to get dialog and find it is null
+  const dialog = document.getElementById("delete-stream-dialog");
+  expect(dialog).toBeNull;
+
+  // expect axios.delete to have been called
+  expect(axios.delete).toHaveBeenCalledWith("http://localhost:8080/stream/1");
```
Comment (ButlerJAL) [inline frontend/src/streamlist/__tests__/StreamList.test.jsx]: Better to use `${process.env.REACT_APP_STREAM}/1`

**Review by ButlerJAL — CHANGES_REQUESTED** [review]
Avoid hard coding URLs. And, could you resolve the few `no-use-before-define` ESLint warnings I am seeing in your tests? Else, we can refactor those later. Thanks!

File: frontend/src/streamlist/__tests__/StreamList.test.jsx [hunk 3]
```diff
@@ -92,3 +94,85 @@ test("Clicking dropdown on table row displays additional information", () => {
...
+  act(() => {
+    ReactTestUtils.Simulate.click(confirmButton);
+  });
+
+  // try to get dialog and find it is null
+  const dialog = document.getElementById("delete-stream-dialog");
+  expect(dialog).toBeNull;
+
+  // expect axios.delete to have been called
+  expect(axios.delete).toHaveBeenCalledWith(`${process.env.REACT_APP_STREAM}/1`);
```
Comment (ButlerJAL) [inline frontend/src/streamlist/__tests__/StreamList.test.jsx]: Reformatting code using Prettier should fix the ESLint warning on this line

**Review by ButlerJAL — DISMISSED** [review]
LGTM.

**Comment by felixlapierre** [comment]
After manual testing w/ the backend on my side, the feature meets the acceptance criteria except for this one, i need to refresh the page for the list to update.

> ...the page in the frontend is updated to show that the stream doesn't exist anymore

Also, the CSS looks a bit wonky on my end 
image
and there is a weird "ripple" effect across the whole screen when the popup opens

Gonna start reviewing the code now

File: frontend/src/api/StreamApi.js [hunk 4]
```diff
@@ -36,3 +36,12 @@ export function getStream(streamId) {
...
   });
 }
+
+export function deleteStream(streamId) {
+  axios
+    .delete(`${process.env.REACT_APP_STREAM}/${streamId}`)
+    // .then(callback)
+    .catch((error) => {
+      // SampleData.deleteStream(streamId, callback);
+    });
```
Comment (felixlapierre) [inline frontend/src/api/StreamApi.js]: There's some commented code here. Also, this function should probably have a callback that gets invoked when no error occurs, so we can refresh the page or remove the deleted stream from the list after a successful delete.

File: frontend/src/streamlist/DeleteStream.jsx [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  const openDeleteDialog = () => {
+    return setOpen(true);
+  };
+  const cancelDelete = () => {
+    return setOpen(false);
+  };
+  const confirmDelete = () => {
+    StreamApi.deleteStream(deleteId);
+    return setOpen(false);
+  }; // call endpoint here, actually
```
Comment (felixlapierre) [inline frontend/src/streamlist/DeleteStream.jsx]: Remove comment

**Review by felixlapierre — DISMISSED** [review]
I have only minor comments! But do look them over and potentially make changes before merging

**Comment by gdzooks** [comment]
> and there is a weird "ripple" effect across the whole screen when the popup opens

i might have skipped to fixing the button's css thing a bit too quickly, but i don't see anything that looks like a ripple? either fixing the button's CSS fixed this too or i'll need a gif/screenshot for a bug report :,)

File: frontend/src/streamlist/DeleteStream.jsx (hunk 5 above)
Comment (gdzooks) [reply frontend/src/streamlist/DeleteStream.jsx]: fixed: [`e51c5ab`](https://github.com/bean-pod/switchboard/pull/211/commits/e51c5ab26b2a0b31e30f721fed2f8ea555d20cfa)

File: frontend/src/api/StreamApi.js (hunk 4 above)
Comment (gdzooks) [reply frontend/src/api/StreamApi.js]: commented code + callback here: [`3d38779`](https://github.com/bean-pod/switchboard/pull/211/commits/3d3877990a2491b1bdfa2d8b6685499f9350b2ac)

**Review by felixlapierre — APPROVED** [review]
CSS issues are fixed, LGTM!!

**Review by ButlerJAL — APPROVED** [review]

**Pull request merged**